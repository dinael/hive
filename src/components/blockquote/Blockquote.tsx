import React, { FC } from 'react'
import styles from './Blockquote.module.scss'

export type BlockquoteProps = {
  title?: string
  footer?: string
  variant: 'fancy' | 'flat'
  className?: string
  children: React.ReactNode
}

export const Blockquote: FC<BlockquoteProps> = ({
  title = '',
  footer = '',
  children,
  variant = 'fancy',
  className = '',
  ...props
}: BlockquoteProps) => {

  const BlockquoteTitle = ({children}:BlockquoteProps) => (
    <p className={styles['blockquote-title']}>
      {children}
    </p>
  )

  const BlockquoteHeader = ({children}:BlockquoteProps) => (
    <header className={styles['blockquote-header']}>
      {children}
    </header>
  )

  const BlockquoteContent = ({children}:BlockquoteProps) => (
    <div className={styles['blockquote-content']}>
      {children}
    </div>
  )

  const BlockquoteFooter = ({children}:BlockquoteProps) => (
    <footer className={styles['blockquote-footer']}>
      {children}
    </footer>
  )

  const BlockquoteItem = () => (
    <blockquote
      className={
        `${styles['blockquote-component']}
          ${styles[variant]}
          ${className ? className : ''}`
      }
      {...props}>
      {title ?
        <BlockquoteHeader variant={'fancy'}>
          {title ?
            <BlockquoteTitle variant={'fancy'}>
              {title}
            </BlockquoteTitle>
          : null}
        </BlockquoteHeader>
      : null}
      {children &&
        <BlockquoteContent variant={'fancy'}>
          {children}
        </BlockquoteContent>}
      {footer &&
        <BlockquoteFooter variant={'fancy'}>
          {footer}
        </BlockquoteFooter>}
    </blockquote>
  )

  return (
    <BlockquoteItem />
  )
}

export default Blockquote
