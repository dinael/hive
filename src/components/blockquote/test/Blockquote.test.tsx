import { describe, test, expect } from 'vitest'
import { render, screen } from '@testing-library/react'
import { Blockquote, BlockquoteProps } from '../Blockquote'

describe('Blockquote', () => {
  const defaultProps: BlockquoteProps = {
    title: 'Blockquote title',
    footer: 'Blockquote footer',
    variant: 'flat',
    className: 'className-test',
    children: 'Blockquote body'
  }

  test('Blockquote renders snapshot', () => {
    const { container } = render(<Blockquote {...defaultProps} />)
    expect(container).toMatchSnapshot()
  })

  test('Blockquote renders correctly', () => {
    const { container } = render(<Blockquote {...defaultProps} />)
    const blockquoteElement = container.querySelector('blockquote')
    expect(blockquoteElement?.tagName).toBe('BLOCKQUOTE')
    expect(container).toMatchSnapshot()
  })

  test('Blockquote renders title correctly', () => {
    const { container } = render(<Blockquote {...defaultProps} />)
    const blockquoteElement = container.querySelector('blockquote > header > .blockquote-title')
    expect(blockquoteElement?.textContent).toBe(defaultProps.title)
    expect(screen.getByText(`${defaultProps.title}`))
    expect(container).toMatchSnapshot()
  })

  test('Blockquote renders quote correctly', () => {
    const { container } = render(<Blockquote {...defaultProps} />)
    const blockquoteElement = container.querySelector('blockquote > .blockquote-content')
    expect(blockquoteElement?.textContent).toBe(defaultProps.children)
    expect(screen.getByText(`${defaultProps.children}`))
    expect(container).toMatchSnapshot()
  })

  test('Blockquote renders footer correctly', () => {
    const { container } = render(<Blockquote {...defaultProps} />)
    const blockquoteElement = container.querySelector('blockquote > footer')
    expect(blockquoteElement?.textContent).toBe(defaultProps.footer)
    expect(screen.getByText(`${defaultProps.footer}`))
    expect(container).toMatchSnapshot()
  })

  test('Blockquote renders with a custom class', () => {
    const { container } = render(<Blockquote {...defaultProps} />)
    const blockquoteElement = container.querySelector('blockquote')
    expect(blockquoteElement?.className).toContain(defaultProps.className)
    expect(container).toMatchSnapshot()
  })
})
