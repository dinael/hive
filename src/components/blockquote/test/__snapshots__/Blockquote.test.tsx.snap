// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Blockquote > renders correctly 1`] = `
<div>
  <blockquote
    class="blockquote-component
          flat
          className-test"
  >
    <header
      class="blockquote-header"
    >
      <p
        class="blockquote-title"
      >
        Blockquote title
      </p>
    </header>
    <div
      class="blockquote-content"
    >
      Blockquote body
    </div>
    <footer
      class="blockquote-footer"
    >
      Blockquote footer
    </footer>
  </blockquote>
</div>
`;

exports[`Blockquote > renders footer correctly 1`] = `
<div>
  <blockquote
    class="blockquote-component
          flat
          className-test"
  >
    <header
      class="blockquote-header"
    >
      <p
        class="blockquote-title"
      >
        Blockquote title
      </p>
    </header>
    <div
      class="blockquote-content"
    >
      Blockquote body
    </div>
    <footer
      class="blockquote-footer"
    >
      Blockquote footer
    </footer>
  </blockquote>
</div>
`;

exports[`Blockquote > renders quote correctly 1`] = `
<div>
  <blockquote
    class="blockquote-component
          flat
          className-test"
  >
    <header
      class="blockquote-header"
    >
      <p
        class="blockquote-title"
      >
        Blockquote title
      </p>
    </header>
    <div
      class="blockquote-content"
    >
      Blockquote body
    </div>
    <footer
      class="blockquote-footer"
    >
      Blockquote footer
    </footer>
  </blockquote>
</div>
`;

exports[`Blockquote > renders snapshot 1`] = `
<div>
  <blockquote
    class="blockquote-component
          flat
          className-test"
  >
    <header
      class="blockquote-header"
    >
      <p
        class="blockquote-title"
      >
        Blockquote title
      </p>
    </header>
    <div
      class="blockquote-content"
    >
      Blockquote body
    </div>
    <footer
      class="blockquote-footer"
    >
      Blockquote footer
    </footer>
  </blockquote>
</div>
`;

exports[`Blockquote > renders title correctly 1`] = `
<div>
  <blockquote
    class="blockquote-component
          flat
          className-test"
  >
    <header
      class="blockquote-header"
    >
      <p
        class="blockquote-title"
      >
        Blockquote title
      </p>
    </header>
    <div
      class="blockquote-content"
    >
      Blockquote body
    </div>
    <footer
      class="blockquote-footer"
    >
      Blockquote footer
    </footer>
  </blockquote>
</div>
`;

exports[`Blockquote > renders with a custom class 1`] = `
<div>
  <blockquote
    class="blockquote-component
          flat
          className-test"
  >
    <header
      class="blockquote-header"
    >
      <p
        class="blockquote-title"
      >
        Blockquote title
      </p>
    </header>
    <div
      class="blockquote-content"
    >
      Blockquote body
    </div>
    <footer
      class="blockquote-footer"
    >
      Blockquote footer
    </footer>
  </blockquote>
</div>
`;
